plaso/parsers/ips_plugins/interface.py

```diff
@@ -0,0 +1,126 @@
+# -*- coding: utf-8 -*-
+"""Interface for IPS log file parser plugins."""
+
+import abc
+import pyparsing
+
+from dfdatetime import time_elements as dfdatetime_time_elements
+
+from plaso.parsers import plugins
+
+
+class IPSPlugin(plugins.BasePlugin):
+  """IPS file parser plugin."""
+
+  NAME = 'ips_plugin'
+  DATA_FORMAT = 'ips log file'
+
+  ENCODING = 'utf-8'
+
+  REQUIRED_HEADER_KEYS = []
+  REQUIRED_CONTENT_KEYS = []
+
+  _TWO_DIGITS = pyparsing.Word(pyparsing.nums, exact=2).set_parse_action(
+      lambda tokens: int(tokens[0], 10))
+
+  _FOUR_DIGITS = pyparsing.Word(pyparsing.nums, exact=4).set_parse_action(
+      lambda tokens: int(tokens[0], 10))
+
+  _VARYING_DIGITS = pyparsing.Word(pyparsing.nums)
+
+  TIMESTAMP_GRAMMAR = (
+      _FOUR_DIGITS.set_results_name('year') + pyparsing.Suppress('-') +
+      _TWO_DIGITS.set_results_name('month') + pyparsing.Suppress('-') +
+      _TWO_DIGITS.set_results_name('day') +
+      _TWO_DIGITS.set_results_name('hours') + pyparsing.Suppress(':') +
+      _TWO_DIGITS.set_results_name('minutes') + pyparsing.Suppress(':') +
+      _TWO_DIGITS.set_results_name('seconds') + pyparsing.Suppress('.') +
+      _VARYING_DIGITS.set_results_name('fraction') +
+      pyparsing.Word(
+          pyparsing.nums + '+' + '-').set_results_name('timezone_delta'))
+
+  def _ParseTimestampValue(self, parser_mediator, timestamp_text):
+    """Parses a timestamp string.
+
+    Args:
+      parser_mediator (ParserMediator): parser mediator.
+      timestamp_text (str): the timestamp to parse.
+
+    Returns:
+       dfdatetime.TimeElements: date and time
+          or None if not available.
+    """
+    # dfDateTime takes the time zone offset as number of minutes relative from
+    # UTC. So for Easter Standard Time (EST), which is UTC-5:00 the sign needs
+    # to be converted, to +300 minutes.
+
+    parsed_timestamp = self.TIMESTAMP_GRAMMAR.parseString(timestamp_text)
+
+    try:
+      time_delta_hours = int(parsed_timestamp['timezone_delta'][:3], 10)
+      time_delta_minutes = int(parsed_timestamp['timezone_delta'][3:], 10)
+    except (TypeError, ValueError):
+      parser_mediator.ProduceExtractionWarning(
+          'unsupported timezone offset value')
+      return None
+
+    time_zone_offset = (time_delta_hours * 60) + time_delta_minutes
+
+    try:
+      fraction_float = float(f"0.{parsed_timestamp['fraction']}")
+      milliseconds = round(fraction_float * 1000)
+
+      time_element_object = dfdatetime_time_elements.TimeElementsInMilliseconds(
```

Comment:
note to self see if TimeElementsWithFractionOfSecond can be used instead

Reply:
Looks like the fraction of second value is either 2 or 4 digits. @rick-slin can you confirm this based on other samples ?

I recommend we preserve the precision here to prevent misrepresentation, also see https://osdfir.blogspot.com/2021/10/pearls-and-pitfalls-of-timeline-analysis.html 

When time permits I'll make some tweaks to dfDateTime to support 10 ms and 100 us intervals

Reply:
I couldn't find an ips file with a timestamp with something other than 10 ms precision (hundredth of seconds).  Where did you encounter one with 100 us precision? I checked on three iPhones (iOS 14, 15, and 16) and a MacBook.

I see your point about preserving granularity.